### PR TITLE
Distributed matrix

### LINKED ISSUE #16: Implement Distributed Matrix of futures of tiles.
See `misc/code.(pdf|pptx)` and `misc/example/m.h` for details.

### PR DESCRIPTION
**Pull request by rasolca**
Extended Matrix class to handle distributed matrices.

Closes #16.

### CODE REVIEW

File: include/dlaf/matrix.h [hunk 1]
```diff
@@ -9,14 +9,24 @@
...
 #include "dlaf/matrix/distribution.h"
 #include "dlaf/matrix/layout_info.h"
 #include "dlaf/tile.h"
 #include "dlaf/types.h"
 
 namespace dlaf {
 
+/// Matrix represent a collection of tiles which contain all the elements of a matrix.
+///
+/// The tiles are distributed according a distribution (see @c Matrix::distribution()),
```
Comment (Sely85) [inline include/dlaf/matrix.h]: typo: "according to a distribution"

File: include/dlaf/matrix.h [hunk 2]
```diff
@@ -25,22 +35,71 @@ class Matrix : public Matrix<const T, device> {
...
+  Matrix(const GlobalElementSize& size, const TileElementSize& block_size,
+         const comm::CommunicatorGrid& comm);
+
+  /// Create a matrix distributed according the distribution @p distribution.
+  Matrix(matrix::Distribution&& distribution);
+
+  /// Create a matrix distributed according the distribution @p distribution,
+  /// specifying the layout.
+  ///
+  /// @param[in] layout is the layout which describe how the elements
```
Comment (Sely85) [inline include/dlaf/matrix.h]: typo: "the layout which describes"

File: include/dlaf/matrix.h [hunk 3]
```diff
@@ -25,22 +35,71 @@ class Matrix : public Matrix<const T, device> {
...
+  ///            of the local part of the matrix will be stored in memory.
+  Matrix(matrix::Distribution&& distribution, const matrix::LayoutInfo& layout);
+
+  /// Create a non distributed matrix,
+  /// which references elements that are already allocated in the memory
+  ///
+  /// @param[in] layout is the layout which describe how the elements
```
Comment (Sely85) [inline include/dlaf/matrix.h]: typo: "is the layout which describes"

File: test/unit/test_matrix.cpp [hunk 4]
```diff
@@ -75,23 +126,61 @@ TYPED_TEST(MatrixTest, Constructor) {
...
+  std::size_t tile_offset = layout.tileOffset(local_tile_index);
+  std::size_t element_offset =
+      sum(tile_element_index.row(), mul(layout.ldTile(), tile_element_index.col()));
   return tile_offset + element_offset;
 }
+/// Returns true if the memory index is stored in distribution.rankIndex().
+bool ownIndex(const Distribution& distribution, const GlobalElementIndex& index) {
+  auto global_tile_index = distribution.globalTileIndex(index);
+  return distribution.rankIndex() == distribution.rankGlobalTile(global_tile_index);
+}
```
Comment (albestro) [inline test/unit/test_matrix.cpp]: Isn't this something that may be useful in general and not just for tests?

File: test/unit/test_matrix.cpp [hunk 5]
```diff
@@ -102,103 +191,209 @@ void checkFromExisting(T* p, const LayoutInfo& layout, Matrix<T, device>& matrix
...
+  // Set the memory elements.
+  // Note: This method is not efficient but for tests is OK.
   for (SizeType j = 0; j < size.cols(); ++j) {
     for (SizeType i = 0; i < size.rows(); ++i) {
-      *ptr({i, j}) = el({i, j});
+      if (own_element({i, j})) {
+        *ptr({i, j}) = el({i, j});
+      }
     }
   }
```
Comment (albestro) [inline test/unit/test_matrix.cpp]: What do you think about extracting a method for this?

File: test/unit/test_matrix.cpp (hunk 4 above)
Comment (rasolca) [reply test/unit/test_matrix.cpp]: No, as the note on line 206 warns, this is the most straightforward but very inefficient method to loop over elements.

File: include/dlaf/matrix.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (albestro) [reply include/dlaf/matrix.h]: Please, check all other similar comments for the same typo.

**Pull request merged**